nwapp/utils/auto-update.js

```diff
@@ -1,31 +1,98 @@
 'use strict';
 
-var log = require('loglevel');
-var notifier = require('./notifier');
-var packageJson = require('../package.json');
 var gui = window.require('nw.gui');
+var manifest = require('../package.json');
+var log = require('loglevel');
+var argv = require('yargs')(gui.App.argv).argv;
 var os = require('./client-type');
-var Updater = require('node-webkit-updater');
+
+var Promise = require('bluebird');
+var urlParse = require('url-parse');
+var rimraf = require('rimraf');
+var path = require('path');
 var temp = require('temp');
 var request = require('request');
 var extract = require('extract-zip');
-var rimraf = require('rimraf');
-var path = require('path');
+var Updater = require('node-webkit-updater');
+
+var notifier = require('./notifier');
+var quitApp = require('./quit-app');
+
+
+// 10 minutes
+var notifyUpdatePollingRate = 10 * 60 * 1000;
+// 24 hours
+var notifyUpdateIntervalTime = 24 * 60 * 60 * 1000;
+var lastNotifyUpdateTime = -1;
+var scheduleUpdateNotifyCallback = function(callback) {
+  var cb = function() {
+    callback();
+    lastNotifyUpdateTime = Date.now();
+  };
+
+  cb();
+  setInterval(function() {
+    var timeSinceNotify = Date.now() - lastNotifyUpdateTime;
+
+    if(timeSinceNotify >= notifyUpdateIntervalTime) {
+      cb();
+    }
+  }, notifyUpdatePollingRate);
+};
+
+
+// You can change the place we use to check and download updates with this CLI parameter `--update-url=192.168.0.58:3010`
+// We use this for testing a release
+var updateUrlOption = argv['update-url'] || 'https://update.gitter.im';
+var transposeUpdateUrl = function(targetUrl) {
+  targetUrl = targetUrl || '';
+  var parsedTargetUrl = urlParse(targetUrl);
+  var parsedUpdateUrl = urlParse(updateUrlOption);
+  parsedTargetUrl.protocol = parsedUpdateUrl.protocol || 'http:';
+  parsedTargetUrl.auth = parsedUpdateUrl.auth;
+  parsedTargetUrl.hostname = parsedUpdateUrl.hostname;
+  parsedTargetUrl.port = parsedUpdateUrl.port;
+
+  var newUrl = parsedTargetUrl.toString();
+  return newUrl;
+};
```

Comment:
I much prefer command line parameters. Easier to manage cross platform and less environment magic.